<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reference counting • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reference counting">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.35.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_shotgun_align.html">Shotgun data - alignments</a>
    </li>
    <li>
      <a href="../articles/04_long_align.html">Long read data - alignments</a>
    </li>
    <li>
      <a href="../articles/05_lineage.html">Lineage calling with SLIMM</a>
    </li>
    <li>
      <a href="../articles/06_counting.html">Reference counting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Reference counting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/06_counting.Rmd"><code>vignettes/06_counting.Rmd</code></a></small>
      <div class="hidden name"><code>06_counting.Rmd</code></div>

    </div>

    
    
<div id="reference-counting" class="section level1">
<h1 class="hasAnchor">
<a href="#reference-counting" class="anchor"></a>Reference counting</h1>
<p>One basic anaylisis we want to perform is to count the abundance of some references in our data. Here reference is a broad term and may refer to transcripts, genomes, or amplicon sequences contained in a reference database.</p>
<p>For this <code>mbtools</code> implements a general purpose expectation maximization (EM) counter. It is implemented in C++ to perform fast and usually reading the alignment file takes longer than the actual analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mbtools)</a></code></pre></div>
<pre><code>## Registered S3 methods overwritten by 'ggplot2':
##   method         from 
##   [.quosures     rlang
##   c.quosures     rlang
##   print.quosures rlang</code></pre>
<pre><code>## Registered S3 method overwritten by 'R.oo':
##   method        from       
##   throw.default R.methodsS3</code></pre>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.12.1
##   - data.table=1.12.2
##   - ggplot2=3.1.1
##   - magrittr=1.5
##   - phyloseq=1.28.0
##   - ShortRead=1.42.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - minimap2=2.15-r905
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<pre><code>## 
## Attaching package: 'mbtools'</code></pre>
<pre><code>## The following object is masked _by_ 'package:BiocGenerics':
## 
##     normalize</code></pre>
<div id="alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#alignment" class="anchor"></a>Alignment</h2>
<p>Let’s start by aligning our short read data to a reference database of 10 microbial genomes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">fi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/shotgun"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="st">      </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>()</a>
<a class="sourceLine" id="cb10-3" title="3">ref &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</a>
<a class="sourceLine" id="cb10-4" title="4">                   <span class="dt">package =</span> <span class="st">"mbtools"</span>)</a>
<a class="sourceLine" id="cb10-5" title="5"></a>
<a class="sourceLine" id="cb10-6" title="6">alns &lt;-<span class="st"> </span><span class="kw"><a href="../reference/align_short_reads.html">align_short_reads</a></span>(</a>
<a class="sourceLine" id="cb10-7" title="7">    fi,</a>
<a class="sourceLine" id="cb10-8" title="8">    <span class="dt">threads =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb10-9" title="9">    <span class="dt">reference =</span> ref,</a>
<a class="sourceLine" id="cb10-10" title="10">    <span class="dt">use_existing =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## INFO [2019-05-25 22:31:01] Aligning 3 samples on 3 threads. Keeping up to 100 secondary alignments.
## INFO [2019-05-25 22:31:03] Finished aligning even1.
## INFO [2019-05-25 22:31:05] Finished aligning even2.
## INFO [2019-05-25 22:31:07] Finished aligning even3.</code></pre>
</div>
<div id="em-counting" class="section level2">
<h2 class="hasAnchor">
<a href="#em-counting" class="anchor"></a>EM counting</h2>
<p>Counting is yet again a workflow and requires an alignment artifact and a configuration.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">config &lt;-<span class="st"> </span><span class="kw"><a href="../reference/config_count.html">config_count</a></span>(</a>
<a class="sourceLine" id="cb12-2" title="2">    <span class="dt">reference =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</a>
<a class="sourceLine" id="cb12-3" title="3">                            <span class="dt">package =</span> <span class="st">"mbtools"</span>),</a>
<a class="sourceLine" id="cb12-4" title="4">    <span class="dt">threads =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb12-5" title="5">    <span class="dt">weights =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb12-6" title="6">)</a>
<a class="sourceLine" id="cb12-7" title="7"></a>
<a class="sourceLine" id="cb12-8" title="8">config</a></code></pre></div>
<pre><code>## $reference
## [1] "/home/cdiener/R/x86_64-pc-linux-gnu-library/3.6/mbtools/extdata/genomes/zymo_mock.fna.gz"
## 
## $threads
## [1] 1
## 
## $method
## [1] "em"
## 
## $maxit
## [1] 1000
## 
## $cutoff
## [1] 0.01
## 
## $tpm
## [1] FALSE
## 
## $weights
## [1] TRUE
## 
## attr(,"class")
## [1] "config"</code></pre>
<p>And we can proceed to counting:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">cn &lt;-<span class="st"> </span><span class="kw"><a href="../reference/count_references.html">count_references</a></span>(alns, config)</a></code></pre></div>
<pre><code>## INFO [2019-05-25 22:31:08] Getting reference lengths from /home/cdiener/R/x86_64-pc-linux-gnu-library/3.6/mbtools/extdata/genomes/zymo_mock.fna.gz...
## INFO [2019-05-25 22:31:08] Normalized IDs. Starting counting...
## INFO [2019-05-25 22:31:09] [alignments/even1.bam] Read 16941 alignments.
## INFO [2019-05-25 22:31:09] [alignments/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-05-25 22:31:09] [alignments/even1.bam] Used 20 EM iterations on 16 equivalence classes. Last max. abs. change was 0.
## INFO [2019-05-25 22:31:09] [alignments/even2.bam] Read 16804 alignments.
## INFO [2019-05-25 22:31:09] [alignments/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-05-25 22:31:09] [alignments/even2.bam] Used 20 EM iterations on 12 equivalence classes. Last max. abs. change was 1.1e-13.
## INFO [2019-05-25 22:31:09] [alignments/even3.bam] Read 17518 alignments.
## INFO [2019-05-25 22:31:09] [alignments/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-05-25 22:31:10] [alignments/even3.bam] Used 20 EM iterations on 20 equivalence classes. Last max. abs. change was 0.</code></pre>
<p>This creates a count artifact which contains the counts.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">cn<span class="op">$</span>counts</a></code></pre></div>
<pre><code>##      reference   counts effective_length sample
##  1: AE004091.2 634.6013          6264306  even1
##  2: AE006468.2 670.7276          4857352  even1
##  3: AE016830.1 661.1689          3217933  even1
##  4: AE017341.1 604.1535         19052864  even1
##  5: AL009126.3 628.9533          4215508  even1
##  6: AL591824.1 629.2075          2944430  even1
##  7: AP008937.1 464.5825          2098587  even1
##  8: BA000007.3 703.2698          5498480  even1
##  9: BK006935.2 584.4928         12072428  even1
## 10: CP000253.1 717.8430          2821263  even1
## 11: AE004091.2 638.9606          6264306  even2
## 12: AE006468.2 547.4441          4857352  even2
## 13: AE016830.1 685.7530          3217933  even2
## 14: AE017341.1 589.4761         19052864  even2
## 15: AL009126.3 588.7936          4215508  even2
## 16: AL591824.1 511.0285          2944430  even2
## 17: AP008937.1 671.5862          2098587  even2
## 18: BA000007.3 631.6744          5498480  even2
## 19: BK006935.2 609.1436         12072428  even2
## 20: CP000253.1 591.1398          2821263  even2
## 21: AE004091.2 617.1627          6264306  even3
## 22: AE006468.2 653.7045          4857352  even3
## 23: AE016830.1 600.9731          3217933  even3
## 24: AE017341.1 639.0245         19052864  even3
## 25: AL009126.3 557.3735          4215508  even3
## 26: AL591824.1 720.0680          2944430  even3
## 27: AP008937.1 732.0089          2098587  even3
## 28: BA000007.3 554.3010          5498480  even3
## 29: BK006935.2 611.6976         12072428  even3
## 30: CP000253.1 486.6862          2821263  even3
##      reference   counts effective_length sample</code></pre>
<p>In each sample we have 10 microbes with the same abundance.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="kw">ggplot</span>(cn<span class="op">$</span>counts, <span class="kw">aes</span>(<span class="dt">x=</span>counts, <span class="dt">y=</span>reference)) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-2" title="2"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">xlim</a></span>(<span class="dv">0</span>, <span class="dv">1000</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-3" title="3"><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sample) <span class="op">+</span><span class="st"> </span><span class="kw">theme_minimal</span>()</a></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Looks ok, but there is a lot of noise since we have very low depth.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw">ggplot</span>(cn<span class="op">$</span>counts, <span class="kw">aes</span>(<span class="dt">x =</span> counts)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">8</span>)</a></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We could also compare that with a naive counting method that just uses the best alignment score. In our case that will perform equally since we have little multi-mapping here.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">config<span class="op">$</span>method &lt;-<span class="st"> "naive"</span></a>
<a class="sourceLine" id="cb20-2" title="2">cn2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/count_references.html">count_references</a></span>(alns, config)</a></code></pre></div>
<pre><code>## INFO [2019-05-25 22:31:10] Getting reference lengths from /home/cdiener/R/x86_64-pc-linux-gnu-library/3.6/mbtools/extdata/genomes/zymo_mock.fna.gz...
## INFO [2019-05-25 22:31:10] Normalized IDs. Starting counting...
## INFO [2019-05-25 22:31:11] [alignments/even1.bam] Read 16941 alignments.
## INFO [2019-05-25 22:31:11] [alignments/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-05-25 22:31:11] [alignments/even2.bam] Read 16804 alignments.
## INFO [2019-05-25 22:31:11] [alignments/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-05-25 22:31:11] [alignments/even3.bam] Read 17518 alignments.
## INFO [2019-05-25 22:31:11] [alignments/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">counts &lt;-<span class="st"> </span>cn<span class="op">$</span>counts[cn2<span class="op">$</span>counts, on =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"sample"</span>, <span class="st">"reference"</span>, <span class="st">"effective_length"</span>)]</a>
<a class="sourceLine" id="cb22-2" title="2"></a>
<a class="sourceLine" id="cb22-3" title="3"><span class="kw">ggplot</span>(counts, <span class="kw">aes</span>(<span class="dt">x =</span> counts, <span class="dt">y =</span> i.counts)) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-4" title="4"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">"lm"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-5" title="5"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"em"</span>, <span class="dt">y =</span> <span class="st">"naive"</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>As we see they both give the same results for that simple case.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#reference-counting">Reference counting</a><ul class="nav nav-pills nav-stacked">
<li><a href="#alignment">Alignment</a></li>
      <li><a href="#em-counting">EM counting</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
