<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Amplicon sequencing analysis • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Amplicon sequencing analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.26.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/amplicon_sequencing.html">Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/preprocessing.html">Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/qa.html">Quality control of sequencing data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Amplicon sequencing analysis</h1>
                        <h4 class="author">Christian Diener</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/amplicon_sequencing.Rmd"><code>vignettes/amplicon_sequencing.Rmd</code></a></small>
      <div class="hidden name"><code>amplicon_sequencing.Rmd</code></div>

    </div>

    
    
<div id="amplicon-sequencing-data-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#amplicon-sequencing-data-analysis" class="anchor"></a>Amplicon sequencing data analysis</h1>
<p>We can combine what we have seen in the quality assessment and preprocessing tutoarials with specific workflow steps for amplicon sequencing, for instance 16S rRNA gene sequencing.</p>
<p>As before we will load <code>mbtools</code> and use our example data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/load_all">load_all</a></span>()</a></code></pre></div>
<pre><code>## Loading mbtools</code></pre>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.10.1
##   - data.table=1.12.0
##   - ggplot2=3.1.0
##   - magrittr=1.5
##   - phyloseq=1.26.1
##   - ShortRead=1.40.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - bowtie2=2.3.4.3
##   - minimap2=2.16-r922
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/example"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>)</a></code></pre></div>
<div id="quality-assessment" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-assessment" class="anchor"></a>Quality assessment</h2>
<p>As before a good first step is to look at the qualities of the raw data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">quals &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>(path) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/quality_control.html">quality_control</a></span>()</a></code></pre></div>
<pre><code>## INFO [2019-03-08 13:14:20] Running quality assay for forward reads from 5 files.
## INFO [2019-03-08 13:14:22] Average per base error is 0.026% (mean score = 35.79).
## INFO [2019-03-08 13:14:22] Mean per cycle entropy is 0.809 (in [0, 2]).
## INFO [2019-03-08 13:14:22] On average we have 8963.80 +- 6059.73 reads per file.
## INFO [2019-03-08 13:14:23] Running quality assay for reverse reads from 5 files.
## INFO [2019-03-08 13:14:25] Average per base error is 0.043% (mean score = 33.69).
## INFO [2019-03-08 13:14:25] Mean per cycle entropy is 0.731 (in [0, 2]).
## INFO [2019-03-08 13:14:25] On average we have 8963.80 +- 6059.73 reads per file.</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">quals<span class="op">$</span>quality_plot</a></code></pre></div>
<pre><code>## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</code></pre>
<p><img src="amplicon_sequencing_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Again we see that we might want to truncate the reads a bit on the 3’ ends to avoid the dip in quality. Depending on your amplified fragment make sure that this leaves sufficient overlap for merging (requires &gt;20bp).</p>
</div>
<div id="denoise" class="section level2">
<h2 class="hasAnchor">
<a href="#denoise" class="anchor"></a>Denoise</h2>
<p>We can now perform our preprocessing and denoise workflow to obtain the amplicon sequence variants (ASVs) in our samples. We can chain our preprocessing and denoise workflows from our original quality assessment. For clarity we will define the confguration for the analysis on top.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">config &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb12-2" title="2">    <span class="dt">preprocess =</span> <span class="kw"><a href="../reference/config_preprocess.html">config_preprocess</a></span>(</a>
<a class="sourceLine" id="cb12-3" title="3">        <span class="dt">trimLeft =</span> <span class="dv">10</span>,  <span class="co"># this is the default</span></a>
<a class="sourceLine" id="cb12-4" title="4">        <span class="dt">trunLen =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">240</span>, <span class="dv">200</span>),  <span class="co"># forward and reverse truncation</span></a>
<a class="sourceLine" id="cb12-5" title="5">        <span class="dt">out_dir =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempdir</a></span>()  <span class="co"># will store filtered files in a temporary dir</span></a>
<a class="sourceLine" id="cb12-6" title="6">    ),</a>
<a class="sourceLine" id="cb12-7" title="7">    <span class="dt">denoise =</span> <span class="kw"><a href="../reference/config_denoise.html">config_denoise</a></span>()  <span class="co"># will only use defaults</span></a>
<a class="sourceLine" id="cb12-8" title="8">)</a>
<a class="sourceLine" id="cb12-9" title="9"></a>
<a class="sourceLine" id="cb12-10" title="10">denoised &lt;-<span class="st"> </span>quals <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</a></code></pre></div>
<pre><code>## INFO [2019-03-08 13:14:27] Preprocessing reads for 5 paired-end samples...
## INFO [2019-03-08 13:14:32] 3.23e+04/4.48e+04 (73.36%) reads passed preprocessing.
## INFO [2019-03-08 13:14:32] Running DADA2 on 1 run(s) from a sample of up to 2.5e+08 bases.
## INFO [2019-03-08 13:14:32] Learning errors for run `all` (5 samples)...
## INFO [2019-03-08 13:14:59] Dereplicating run `all` (5 samples)...
## INFO [2019-03-08 13:15:02] Inferring sequence variants for run `all`...
## INFO [2019-03-08 13:15:12] Finished run `all`.
## INFO [2019-03-08 13:15:12] Merged sequence tables, now removing chimeras...
## INFO [2019-03-08 13:15:12] Removed 5/164 sequence variants as chimeric (1.20% of reads)
## INFO [2019-03-08 13:15:12] Assigning taxonomy to 159 sequences...
## INFO [2019-03-08 13:15:12] Downloading taxa db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpcqOKTY/taxa.fna.gz...
## INFO [2019-03-08 13:16:29] Downloading taxa db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpcqOKTY/species.fna.gz...
## INFO [2019-03-08 13:21:00] Classified variants: Kingdom = 100%, Phylum = 100%, Class = 100%, Order = 100%, Family = 98.7%, Genus = 66.7%, Species = 8.2%
## INFO [2019-03-08 13:21:00] Hashing 159 sequence variants.</code></pre>
<p>This will run both workflow steps sequentially and will use all available CPUs by default. You will get some disgnostic output on the logging interface but you can also inspect those in the returned artifact.</p>
<p>For instance to see how many reads were conserved in each processing step:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">denoised<span class="op">$</span>passed_reads</a></code></pre></div>
<p>We see that most samples were conserved (as reported in the logging).</p>
<p>The abundance of each sequence variant is saved in <code>denoised$feature_tables</code> which contains abundances for samples x ASVs. The ASVs here have been named by their md5 hash value and you can get the taxonomy and original sequence from <code>denoised$taxonomy</code>:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">denoised<span class="op">$</span>taxonomy[<span class="dv">1</span>, ]</a></code></pre></div>
<pre><code>##                                                                                                                                                                                                                                                      Kingdom 
##                                                                                                                                                                                                                                                   "Bacteria" 
##                                                                                                                                                                                                                                                       Phylum 
##                                                                                                                                                                                                                                              "Bacteroidetes" 
##                                                                                                                                                                                                                                                        Class 
##                                                                                                                                                                                                                                                "Bacteroidia" 
##                                                                                                                                                                                                                                                        Order 
##                                                                                                                                                                                                                                              "Bacteroidales" 
##                                                                                                                                                                                                                                                       Family 
##                                                                                                                                                                                                                                             "Muribaculaceae" 
##                                                                                                                                                                                                                                                        Genus 
##                                                                                                                                                                                                                                                           NA 
##                                                                                                                                                                                                                                                      Species 
##                                                                                                                                                                                                                                                           NA 
##                                                                                                                                                                                                                                                     sequence 
## "ACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGAAGATCAAGTCAGCGGTAAAATTGAGAGGCTCAACCTCTTCGAGCCGTTGAAACTGGTTTTCTTGAGTGAGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCTCAACTGACGCTCATGCACGAAAGTGTGGGTATCGAACAG"</code></pre>
<p>To see all steps run to obatin that output you can check the provenance:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">denoised<span class="op">$</span>steps</a></code></pre></div>
<pre><code>## [1] "quality_control" "preprocess"      "denoise"</code></pre>
</div>
<div id="managing-several-sequencing-runs" class="section level2">
<h2 class="hasAnchor">
<a href="#managing-several-sequencing-runs" class="anchor"></a>Managing several sequencing runs</h2>
<p>The <code>denoise</code> workflow step supports multiple sequencing runs in the same experiment. This requires that your files data table has a “run” column. If you sequencing files are organized such that every subfolder is a run you can simply use <code><a href="../reference/find_read_files.html">find_read_files(path, dirs_are_runs = TRUE)</a></code>. Otherwise, you have to specify the run by hand. We will do so for our files here:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">files &lt;-<span class="st"> </span>quals<span class="op">$</span>files</a>
<a class="sourceLine" id="cb19-2" title="2">files[, run <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">times =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>))]</a>
<a class="sourceLine" id="cb19-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(files)</a></code></pre></div>
<pre><code>##                                                                            forward
## 1: /Users/cdiener/code/mbtools/inst/extdata/example/F3D0_S188_L001_R1_001.fastq.gz
## 2: /Users/cdiener/code/mbtools/inst/extdata/example/F3D1_S189_L001_R1_001.fastq.gz
## 3: /Users/cdiener/code/mbtools/inst/extdata/example/F3D2_S190_L001_R1_001.fastq.gz
## 4: /Users/cdiener/code/mbtools/inst/extdata/example/F3D3_S191_L001_R1_001.fastq.gz
## 5: /Users/cdiener/code/mbtools/inst/extdata/example/Mock_S280_L001_R1_001.fastq.gz
##                                                                            reverse
## 1: /Users/cdiener/code/mbtools/inst/extdata/example/F3D0_S188_L001_R2_001.fastq.gz
## 2: /Users/cdiener/code/mbtools/inst/extdata/example/F3D1_S189_L001_R2_001.fastq.gz
## 3: /Users/cdiener/code/mbtools/inst/extdata/example/F3D2_S190_L001_R2_001.fastq.gz
## 4: /Users/cdiener/code/mbtools/inst/extdata/example/F3D3_S191_L001_R2_001.fastq.gz
## 5: /Users/cdiener/code/mbtools/inst/extdata/example/Mock_S280_L001_R2_001.fastq.gz
##      id injection_order lane run
## 1: F3D0             188    1   1
## 2: F3D1             189    1   1
## 3: F3D2             190    1   2
## 4: F3D3             191    1   2
## 5: Mock             280    1   3</code></pre>
<p>Rerunning the previous workflow with this file list will now give a slightly different output:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/load_all">load_all</a></span>(<span class="st">"~/code/mbtools"</span>)</a></code></pre></div>
<pre><code>## Loading mbtools</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1">denoised &lt;-<span class="st"> </span>files <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb23-2" title="2"><span class="st">            </span><span class="kw"><a href="../reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</a></code></pre></div>
<pre><code>## INFO [2019-03-08 13:21:03] Preprocessing reads for 5 paired-end samples...
## INFO [2019-03-08 13:21:08] 3.23e+04/4.48e+04 (73.36%) reads passed preprocessing.
## INFO [2019-03-08 13:21:08] Running DADA2 on 3 run(s) from a sample of up to 2.5e+08 bases.
## INFO [2019-03-08 13:21:08] Learning errors for run `1` (2 samples)...
## INFO [2019-03-08 13:21:21] Dereplicating run `1` (2 samples)...
## INFO [2019-03-08 13:21:22] Inferring sequence variants for run `1`...
## INFO [2019-03-08 13:21:26] Finished run `1`.
## INFO [2019-03-08 13:21:26] Learning errors for run `2` (2 samples)...
## INFO [2019-03-08 13:21:42] Dereplicating run `2` (2 samples)...
## INFO [2019-03-08 13:21:43] Inferring sequence variants for run `2`...
## INFO [2019-03-08 13:21:49] Finished run `2`.
## INFO [2019-03-08 13:21:49] Learning errors for run `3` (1 samples)...
## INFO [2019-03-08 13:21:50] Dereplicating run `3` (1 samples)...
## INFO [2019-03-08 13:21:50] Inferring sequence variants for run `3`...
## INFO [2019-03-08 13:21:51] Finished run `3`.
## INFO [2019-03-08 13:21:51] Merged sequence tables, now removing chimeras...
## INFO [2019-03-08 13:21:51] Removed 5/162 sequence variants as chimeric (1.20% of reads)
## INFO [2019-03-08 13:21:51] Assigning taxonomy to 157 sequences...
## INFO [2019-03-08 13:21:51] Downloading taxa db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpcqOKTY/taxa.fna.gz...
## INFO [2019-03-08 13:23:41] Downloading taxa db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpcqOKTY/species.fna.gz...
## INFO [2019-03-08 13:27:21] Classified variants: Kingdom = 100%, Phylum = 100%, Class = 100%, Order = 100%, Family = 98.7%, Genus = 65.6%, Species = 8.3%
## INFO [2019-03-08 13:27:21] Hashing 157 sequence variants.</code></pre>
<p>Error profiles will now be estimated for each run individually and final tables will be merged at the end. This is slower than having all files in one run so you should always try to distribute your samples across as few sequencing runs as possible.</p>
</div>
<div id="converting-to-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-to-phyloseq" class="anchor"></a>Converting to phyloseq</h2>
<p>For all further analyses you might want to work with a data type which is more suited for amplicon abundance data. You can directly obtain a phyloseq object with</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">ps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_phyloseq.html">as_phyloseq</a></span>(denoised)</a></code></pre></div>
<p>It is possible to pass in metadata as a data frame with <code>to_phyloseq(object, metadata = df)</code>.</p>
<p>You can now use all additional functions from <code>mbtools</code> on that. For instance we can get the ASV counts in a tidy data table:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1">asv_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/taxa_count.html">taxa_count</a></span>(ps, <span class="dt">lev =</span> <span class="ot">NA</span>)</a>
<a class="sourceLine" id="cb26-2" title="2">asv_counts[sample <span class="op">==</span><span class="st"> "Mock"</span>]</a></code></pre></div>
<pre><code>##      sample                             taxa reads          species
##   1:   Mock c69a3db92e1dcdebb01dfaa5b5795559     0             &lt;NA&gt;
##   2:   Mock c6d91507bb3825befa2f3aa9c7631ac5     0             &lt;NA&gt;
##   3:   Mock 81a461aaeb070ec0d623dd828322a7ad     0     Alistipes NA
##   4:   Mock 0a98021c7909137422f0417bdf82250e     0             &lt;NA&gt;
##   5:   Mock adcedd4ffc4b4d4554d920a7ea1ab935     0             &lt;NA&gt;
##  ---                                                               
## 153:   Mock e706eefe517d197c30cc503de74f7741     0 GCA-900066225 NA
## 154:   Mock 5d4e5c3d9eee3918b6d4a300e4ed3e6c     0             &lt;NA&gt;
## 155:   Mock d49f2ec46e108728c7f0e387ab139336     0             &lt;NA&gt;
## 156:   Mock 31fa7eed69f8fd4b4e65076051f4e044     0             &lt;NA&gt;
## 157:   Mock 4233dee7fda3b32a9f712023ded9973b     0             &lt;NA&gt;</code></pre>
<p>Here we see that we have identified that there are 20 ASVs in the Mock sample which is the exact number of strains in that sample.</p>
<p>We could also plot the phyla distribution across the samples:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="kw"><a href="../reference/plot_taxa.html">plot_taxa</a></span>(ps)</a></code></pre></div>
<pre><code>## Warning in `[.data.table`(counts, , `:=`(id, as.numeric(sample))):
## Invalid .internal.selfref detected and fixed by taking a (shallow) copy
## of the data.table so that := can add this new column by reference. At
## an earlier point, this data.table has been copied by R (or was created
## manually using structure() or similar). Avoid key&lt;-, names&lt;- and attr&lt;-
## which in R currently (and oddly) may copy the whole data.table. Use set*
## syntax instead to avoid copying: ?set, ?setnames and ?setattr. If this
## message doesn't help, please report your use case to the data.table issue
## tracker so the root cause can be fixed or this message improved.</code></pre>
<p><img src="amplicon_sequencing_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#amplicon-sequencing-data-analysis">Amplicon sequencing data analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#quality-assessment">Quality assessment</a></li>
      <li><a href="#denoise">Denoise</a></li>
      <li><a href="#managing-several-sequencing-runs">Managing several sequencing runs</a></li>
      <li><a href="#converting-to-phyloseq">Converting to phyloseq</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
